File: apps/pos/native-windows/DesktopKitchenPOS/ViewModels/KitchenViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using DesktopKitchenPOS.Models;
using DesktopKitchenPOS.Utilities;

namespace DesktopKitchenPOS.ViewModels;

public partial class KitchenViewModel : ObservableObject
{
    [ObservableProperty]
    private List<Order> _orders = new();

    [ObservableProperty]
    private bool _isLoading = true;

    [ObservableProperty]
    private string? _error;

    private PeriodicTimer? _pollingTimer;
    private PeriodicTimer? _clockTimer;
    private CancellationTokenSource? _cts;

    public int PendingCount => Orders.Count(o => o.Status == "pending");

    public void StartPolling()
    {
        StopPolling();
        _cts = new CancellationTokenSource();

        _ = PollOrdersAsync(_cts.Token);
        _ = PollClockAsync(_cts.Token);
    }

    public void StopPolling()
    {
        _cts?.Cancel();
        _pollingTimer?.Dispose();
        _clockTimer?.Dispose();
    }

    private async Task PollOrdersAsync(CancellationToken ct)
    {
        await FetchOrdersAsync();
        _pollingTimer = new PeriodicTimer(TimeSpan.FromSeconds(5));
        while (await _pollingTimer.WaitForNextTickAsync(ct))
        {
            await FetchOrdersAsync();
        }
    }

    private async Task PollClockAsync(CancellationToken ct)
    {
        _clockTimer = new PeriodicTimer(TimeSpan.FromSeconds(1));
        while (await _clockTimer.WaitForNextTickAsync(ct))
        {
            // Force UI to refresh elapsed times
            OnPropertyChanged(nameof(Orders));
        }
    }

    private async Task FetchOrdersAsync()
    {
        try
        {
            var data = await DesktopKitchenPOS.App.OrderService.GetKitchenOrdersAsync();
            var statusRank = new Dictionary<string, int>
            {
                ["pending"] = 0, ["preparing"] = 1, ["confirmed"] = 2
            };
            Orders = data
                .Where(o => o.Status is not "completed" and not "cancelled")
                .OrderBy(o => statusRank.GetValueOrDefault(o.Status, 3))
                .ThenByDescending(o => o.CreatedAt ?? "")
                .ToList();
            Error = null;
            OnPropertyChanged(nameof(PendingCount));
        }
        catch (Exception ex)
        {
            Error = ex.Message;
        }
        IsLoading = false;
    }

    [RelayCommand]
    public async Task StartOrderAsync(int id)
    {
        try
        {
            await DesktopKitchenPOS.App.OrderService.UpdateStatusAsync(id, "preparing");
            await FetchOrdersAsync();
        }
        catch (Exception ex)
        {
            Error = ex.Message;
        }
    }

    [RelayCommand]
    public async Task MarkReadyAsync(int id)
    {
        try
        {
            await DesktopKitchenPOS.App.OrderService.UpdateStatusAsync(id, "ready");
            await FetchOrdersAsync();
        }
        catch (Exception ex)
        {
            Error = ex.Message;
        }
    }

    public int ElapsedSeconds(Order order) => DateFormatters.ElapsedSeconds(order.CreatedAt);

    public bool IsUrgent(Order order) => ElapsedSeconds(order) > 600;
}

File: apps/pos/native-android/app/src/main/java/com/desktopkitchen/pos/viewmodels/KitchenViewModel.kt
package com.desktopkitchen.pos.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.desktopkitchen.pos.models.Order
import com.desktopkitchen.pos.services.OrderService
import com.desktopkitchen.pos.utilities.DateFormatters
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class KitchenViewModel @Inject constructor(
    private val orderService: OrderService
) : ViewModel() {

    var orders by mutableStateOf<List<Order>>(emptyList())
        private set
    var isLoading by mutableStateOf(true)
        private set
    var error by mutableStateOf<String?>(null)
        private set
    var currentTimeMillis by mutableLongStateOf(System.currentTimeMillis())
        private set

    private var pollingJob: Job? = null
    private var timerJob: Job? = null

    val pendingCount: Int
        get() = orders.count { it.status == "pending" }

    fun startPolling() {
        pollingJob?.cancel()
        pollingJob = viewModelScope.launch {
            fetchOrders()
            while (isActive) {
                delay(5000)
                fetchOrders()
            }
        }

        timerJob?.cancel()
        timerJob = viewModelScope.launch {
            while (isActive) {
                delay(1000)
                currentTimeMillis = System.currentTimeMillis()
            }
        }
    }

    fun stopPolling() {
        pollingJob?.cancel()
        timerJob?.cancel()
    }

    private suspend fun fetchOrders() {
        try {
            val data = orderService.getKitchenOrders()
            val statusRank = mapOf("pending" to 0, "preparing" to 1, "confirmed" to 2)
            orders = data
                .filter { it.status != "completed" && it.status != "cancelled" }
                .sortedWith(compareBy<Order> { statusRank[it.status] ?: 3 }
                    .thenByDescending { it.created_at ?: "" })
            error = null
        } catch (e: Exception) {
            error = e.message
        }
        isLoading = false
    }

    fun startOrder(id: Int) {
        viewModelScope.launch {
            try {
                orderService.updateStatus(id, "preparing")
                fetchOrders()
            } catch (e: Exception) {
                error = e.message
            }
        }
    }

    fun markReady(id: Int) {
        viewModelScope.launch {
            try {
                orderService.updateStatus(id, "ready")
                fetchOrders()
            } catch (e: Exception) {
                error = e.message
            }
        }
    }

    fun elapsedSeconds(order: Order): Int {
        val created = order.created_at ?: return 0
        return DateFormatters.elapsedSeconds(created)
    }

    fun isUrgent(order: Order): Boolean = elapsedSeconds(order) > 600

    override fun onCleared() {
        super.onCleared()
        stopPolling()
    }
}
